theano/sandbox/rng_mrg.py

```diff
@@ -1117,6 +1118,79 @@ def normal(self, size, avg=0.0, std=1.0, ndim=None,
         assert final_samples.dtype == dtype
         return final_samples
 
+    def truncated_normal(self, size, avg=0.0, std=1.0, ndim=None,
+                         dtype=None, nstreams=None):
+        # TODO : need description for ndim
+        """
+        Sample a tensor of values from a normal distribution
+        truncated at 2 standard deviations from the mean.
+
+        Parameters
+        ----------
+        avg
+            The mean value for the truncated normal to sample from.
+        std
+            The standard deviation for the truncated normal to sample from.
+        size
+            Can be a list of integers or Theano variables (ex: the shape
+            of another Theano Variable).
+        dtype
+            The output data type. If dtype is not specified, it will be
+            inferred from the dtype of low and high, but will be at
+            least as precise as floatX.
+        nstreams
+            Number of streams.
+
+        """
+        avg = undefined_grad(as_tensor_variable(avg))
+        std = undefined_grad(as_tensor_variable(std))
+
+        if dtype is None:
+            dtype = scal.upcast(config.floatX, avg.dtype, std.dtype)
+
+        size = cast(size, 'int64')
+        avg = cast(avg, dtype)
+        std = cast(std, dtype)
+
+        # generate even number of uniform samples
+        n_odd_samples = prod(size)
+        n_even_samples = n_odd_samples + n_odd_samples % 2
+        uniform = self.uniform(size=(n_even_samples, ), low=0., high=1.,
+                               ndim=1, dtype=dtype, nstreams=nstreams)
+
+        # box-muller transform
+        u1 = uniform[:n_even_samples // 2]
+        u2 = uniform[n_even_samples // 2:]
+        r = sqrt(-2.0 * log(u1))
+        theta = np.array(2.0 * np.pi, dtype=dtype) * u2
+        cos_theta, sin_theta = cos(theta), sin(theta)
+        z0 = r * cos_theta
+        z1 = r * sin_theta
+
+        # use valid samples
+        to_fix0 = bitwise_or(z0 < -2., z0 > 2.)
+        to_fix1 = bitwise_or(z1 < -2., z1 > 2.)
+        z0_valid = z0[nonzero(bitwise_not(to_fix0))]
+        z1_valid = z1[nonzero(bitwise_not(to_fix1))]
+
+        # re-sample invalid samples
+        to_fix0 = nonzero(to_fix0)[0]
+        to_fix1 = nonzero(to_fix1)[0]
+        n_fix_samples = to_fix0.size + to_fix1.size
+        u_fix = self.uniform(size=(n_fix_samples, ), low=0.13585, high=1.,
```

Comment:
From the wikipedia page, I think the "low" number would be dtype dependent.

I would be interrested to know how the other framework does it. Did you looked at that? I checked tensorflow and it a variable number of time until it accept a number. The way you do it seem interresting and could prevent that need.

I would like the input from @lamblin and @abergeron on the way you implement it. resample only once with with a different uniform to force a max of 2 stddev.

Reply:
In what sense would `low` depend on the `dtype`? If `low = 0.`, the `dtype` will indeed cause tail-truncation (that is where my idea for the resampling came from).

Which other framework? Keras simply samples from a normal and clips all values to the edges, leaving a distribution that is very non-Gaussian and severely messes with the variance. Tensorflow does indeed some iterative resampling and I have no idea where the CNTK implementation is hidden.